libff/algebra/fields/binary_field.hpp

```diff
@@ -0,0 +1,101 @@
+/** @file
+ *****************************************************************************
+ Declaration of common API for all finite fields in the binary/ directory.
+
+ Currently NOT used by the fields in this library. This class is not actually
+ the parent class of any field. All APIs are enforced through tests instead.
+
+ The reason for this is to ensure high performance of all fields. This class
+ exists as documentation for common API between fields.
+
+ Includes fields F2^n for specified n. All of the binary entension fields must
+ implement all functions declared in this class.
+ *****************************************************************************
+ * @author     This file is part of libff, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+#include <libff/algebra/field_utils/bigint.hpp>
+#include <vector>
+
+namespace libff {
+
+template<typename T>
+class BinaryField;
+
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class gf32 : public BinaryField<gf32> ...
+ */
+template<typename T>
+class BinaryField {
+public:
+    /* Functions unique to binary fields */
+
+    /** Returns the constituent bits in 64 bit words, in little-endian order. */
+    std::vector<uint64_t> as_words() const = 0;
+
+    static const constexpr uint64_t modulus_;
+    static const constexpr uint64_t num_bits;
+
+    /** generator of gf2^n */
+    static T multiplicative_generator;
+
+    /** If extension field, returns the base field's characteristic. */
+    template<mp_size_t n>
+    static constexpr bigint<n> field_char() { return bigint<n>(2); }
+
+    /* Functions common to all finite fields */
+#ifdef PROFILE_OP_COUNTS // NOTE: op counts are affected when you exponentiate with ^
+    static long long add_cnt;
+    static long long sub_cnt;
+    static long long mul_cnt;
+    static long long sqr_cnt;
+    static long long inv_cnt;
+#endif
+
+    virtual T& operator+=(const T& other) = 0;
```

Comment:
Is it okay for these to be virtual, will we lose efficiency due to vtable issues?